### DISCUSSION TITLE
plugin for intellij idea

### DISCUSSION THREAD

**Discussion by MohammadMD1383** [vlang/v#14922] [Ideas]
I started working on a plugin for intellij based ides.

https://github.com/MohammadMD1383/IntelliJ-V

but I need the help of community. (for complete grammar rules, etc ...)

I need your help guys!

---

screenshot(partial and incomplete syntax highlighting for now):
![image](https://user-images.githubusercontent.com/69088224/176999573-e4f1d9fe-1d22-4297-83bb-6ad3a479ef2d.png)

**Comment by JalonSolov** [comment]
https://github.com/nedpals/tree-sitter-v may help

**Comment by i582** [comment]
Funny enough, I started making a prototype on the day this discussion was created :)

I am now parsing a fairly large subset of the language, but it is still far from complete, since the grammar has not yet been completed.

For example file from builtin folder:

<img width="553" src="https://user-images.githubusercontent.com/51853996/177897027-b5dca7b3-475f-412c-ba18-d40493b1be48.png">

See repo: https://github.com/i582/vlang-idea

Other features include "New file action", partially Go To Symbol, built-in vfmt that works as standard IDEA formatting, auto-closing of quotes and brackets, commenting with IDEA shortcuts (including partial support for Doc comments), auto-completion of function names, structures, enums, etc. (but not methods yet). There is also a simple run configuration, but there are still problems with it, but in general, now you can run any `main()` through the icon next to the function. Also plugin highlights the function name, types, etc.

The plugin is still far from being released. but I think we can team up to make the plugin faster 😄

**Reply by MohammadMD1383** [reply]
great! so I will delete my own repo and will try to contribute to yours

**Reply by i582** [reply]
I think it’s worth making some roadmap, I still don’t know where this is going, but I want to believe that with the support of the community I can finish it, at least at the parsing level.

**Reply by i582** [reply]
According to discord, there have already been at least two attempts to make a plugin for idea, but something seems to have gone wrong 😄

**Reply by MohammadMD1383** [reply]
well, I think completing the syntax highlighting feature could be the first task to complete

**Reply by i582** [reply]
yep, without this it will be very difficult to talk about the usefulness of the plugin

**Reply by MohammadMD1383** [reply]
> According to discord, there have already been at least two attempts to make a plugin for idea, but something seems to have gone wrong smile

but we do it 💪

**Reply by i582** [reply]
V is becoming more and more stable, which is good news. perhaps it would be rather strange to make a plugin earlier, since a lot of things changed in the language

**Reply by MohammadMD1383** [reply]
well, I forked your project

**Reply by medvednikov** [reply]
By the way, VLS is a lot more stable now, it works with Sublime and neovim, so in theory it should work with IDEA as well (with the IDEA LS plugin).

**Reply by MohammadMD1383** [reply]
> By the way, VLS is a lot more stable now, it works with Sublime and neovim, so in theory it should work with IDEA as well (with the IDEA LS plugin).

I tried it in vscode, a few days ago, but that was awful.

**Reply by i582** [reply]
> By the way, VLS is a lot more stable now, it works with Sublime and neovim, so in theory it should work with IDEA as well (with the IDEA LS plugin).

In theory :)
But as far as I know the idea plugin for LSP is pretty bad and is unstable and slow.

**Reply by i582** [reply]
Also, the plugin for the IDE is not only language support, but also support for projects, modules, builds and debugging. I think in idea it is possible to make a much more complex and convenient system, which will undoubtedly increase the number of V users.

**Reply by medvednikov** [reply]
> idea plugin for LSP is pretty bad and is unstable and slow

that's unfortunate
LS is a great way to standardize this across all IDEs

> Also, the plugin for the IDE is not only language support

all right, thanks, and good luck!

I'll add it to the home page next to the vscode plugin once it's stable

**Reply by i582** [reply]
yep, LSP is a great technology, it would be much easier if we only had to do project support, launch configurations, and so on.

**Reply by medvednikov** [reply]
@i582 Ned has also been working on this:

https://github.com/nedpals/intellij-vlang

It already works somewhat:

https://discord.com/channels/592103645835821068/784039809459027979/996428951121571921

**Reply by i582** [reply]
> @i582 Ned has also been working on this:
> 
> https://github.com/nedpals/intellij-vlang
> 
> It already works somewhat:
> 
> https://discord.com/channels/592103645835821068/784039809459027979/996428951121571921

Oh interesting, I took a break from plugin development for a while and didn't notice.

**Comment by i582** [comment]
And finally we here: https://intellij-v.github.io

**Reply by MohammadMD1383** [reply]
I was really looking forward to seeing this. 🙏👌😃
